src/v/redpanda/admin_server.cc

```diff
@@ -3435,6 +3435,103 @@ void admin_server::register_usage_routes() {
+        seastar::httpd::debug_json::raft_follower_state metrics;
+        metrics.id = follower_stat.node_id.id();
+        metrics.last_flushed_log_index = follower_stat.last_flushed_log_index;
+        metrics.last_dirty_log_index = follower_stat.last_dirty_log_index;
+        metrics.match_index = follower_stat.match_index;
+        metrics.next_index = follower_stat.next_index;
+        metrics.last_sent_offset = follower_stat.last_sent_offset;
+        metrics.heartbeats_failed = follower_stat.heartbeats_failed;
+        metrics.is_learner = follower_stat.is_learner;
+        metrics.ms_since_last_heartbeat
+          = std::chrono::duration_cast<std::chrono::milliseconds>(
+              raft::clock_type::now()
+              - follower_stat.last_received_reply_timestamp)
+              .count();
+        metrics.last_sent_seq = follower_stat.last_sent_seq;
+        metrics.last_received_seq = follower_stat.last_received_seq;
+        metrics.last_successful_received_seq
+          = follower_stat.last_successful_received_seq;
+        metrics.suppress_heartbeats = static_cast<bool>(
+          follower_stat.suppress_heartbeats);
+        metrics.is_recovering = follower_stat.is_recovering;
+        ans.followers.push(metrics);
+    }
+
+    return ans;
+}
+
+} // namespace
+
+ss::future<ss::json::json_return_type>
+admin_server::get_raft_state_handler(std::unique_ptr<ss::httpd::request> req) {
+    const model::ntp ntp = parse_ntp_from_request(req->param);
+
+    if (need_redirect_to_leader(ntp, _metadata_cache)) {
+        throw co_await redirect_to_leader(*req, ntp);
+    }
+
+    const bool is_controller = ntp == model::controller_ntp;
+    cluster::consensus_ptr raft;
+
+    if (is_controller) {
+        co_return co_await ss::smp::submit_to(
+          ss::shard_id(0),
+          [ntp, this]() -> ss::future<ss::json::json_return_type> {
+              return ss::make_ready_future<ss::json::json_return_type>(
+                fill_raft_status_for_ntp(_controller->get_raft0(), ntp));
+          });
+    } else {
+        auto shard = _shard_table.local().shard_for(ntp);
+        if (!shard) {
+            throw ss::httpd::not_found_exception(
+              fmt::format("Could not find shard for ntp: {}", ntp));
+        }
+        co_return co_await _partition_manager.invoke_on(
+          shard.value(),
+          [ntp](auto& pm) -> ss::future<ss::json::json_return_type> {
+              auto partition = pm.get(ntp);
+              if (!partition) {
+                  auto ex = ss::httpd::base_exception(
+                    fmt::format("Can not find partition object for {}", ntp),
+                    ss::httpd::reply::status_type::service_unavailable);
+                  return ss::make_exception_future<ss::json::json_return_type>(
+                    ex);
+              }
+              return ss::make_ready_future<ss::json::json_return_type>(
+                fill_raft_status_for_ntp(partition->raft(), ntp));
```

Comment:
>                 fill_raft_status_for_ntp(partition->raft(), ntp));

that throws directly. please return an exceptional future if in a non-coroutine context.